### ISSUE TITLE
Beta Code: canonicalize order of multiple diacritics

### ISSUE THREAD

**Issue by whoopsedesy** [sasansom/sedes#45]
#38 fixed many issues with the rendering of multiple combining diacritics, but there are still some problems, for example in the first word of [Sh. 5](https://sasansom.github.io/sedes/shield.html#book-Sh.-line-5):

![sh-5](https://user-images.githubusercontent.com/53617134/109072971-a99e9080-76ed-11eb-84a7-43ee47bbcd2f.png)

While the [same word in Perseus](http://www.perseus.tufts.edu/hopper/text?doc=Perseus:text:1999.01.0127) looks fine:

![perseus-sh-5](https://user-images.githubusercontent.com/53617134/109073026-bf13ba80-76ed-11eb-843e-66972b2ebc78.png)

The issue comes down to the order of diacritics in Beta Code (which map to combining characters in Unicode). ΐ needs to be represented as ϊ plus tonos, not as ί plus dialytika.

The [Beta Code for this word](https://github.com/sasansom/sedes/blob/d3282b0da8e76482dc5c681c2fd37f50d4136c22/corpus/shield.xml#L87) is `ei)/dei/+`, where
* `)` [maps to](https://stephanus.tlg.uci.edu/encoding/BCM.pdf#page=6) U+0313 COMBINING COMMA ABOVE
* `/` maps to U+0301 COMBINING ACUTE ACCENT
* `+` maps to U+0308 COMBINING DIAERESIS

Decoding the Beta Code into NFD Unicode, while keeping diacritics in the same order, yields
```
U+03B5 GREEK SMALL LETTER EPSILON
U+03B9 GREEK SMALL LETTER IOTA
U+0313 COMBINING COMMA ABOVE
U+0301 COMBINING ACUTE ACCENT
U+03B4 GREEK SMALL LETTER DELTA
U+03B5 GREEK SMALL LETTER EPSILON
U+03B9 GREEK SMALL LETTER IOTA
U+0301 COMBINING ACUTE ACCENT
U+0308 COMBINING DIAERESIS
```

However, when these characters are composed into NFC, the U+0301 acute accent combines with the preceding iota to form GREEK SMALL LETTER IOTA WITH TONOS, but the U+0308 remains as a combining character:
```
U+03B5 GREEK SMALL LETTER EPSILON
U+1F34 GREEK SMALL LETTER IOTA WITH PSILI AND OXIA
U+03B4 GREEK SMALL LETTER DELTA
U+03B5 GREEK SMALL LETTER EPSILON
U+03AF GREEK SMALL LETTER IOTA WITH TONOS
U+0308 COMBINING DIAERESIS
```

If, instead, we swap the order of U+0301 and U+0308 to put the diaeresis first,
```
U+03B5 GREEK SMALL LETTER EPSILON
U+03B9 GREEK SMALL LETTER IOTA
U+0313 COMBINING COMMA ABOVE
U+0301 COMBINING ACUTE ACCENT
U+03B4 GREEK SMALL LETTER DELTA
U+03B5 GREEK SMALL LETTER EPSILON
U+03B9 GREEK SMALL LETTER IOTA
U+0308 COMBINING DIAERESIS
U+0301 COMBINING ACUTE ACCENT
```

The NFC combines both diacritics with the iota to form GREEK SMALL LETTER IOTA WITH DIALYTIKA AND TONOS, which renders properly and is what Perseus has:
```
U+03B5 GREEK SMALL LETTER EPSILON
U+1F34 GREEK SMALL LETTER IOTA WITH PSILI AND OXIA
U+03B4 GREEK SMALL LETTER DELTA
U+03B5 GREEK SMALL LETTER EPSILON
U+0390 GREEK SMALL LETTER IOTA WITH DIALYTIKA AND TONOS
```

Indeed, if you configure the Perseus display preferences to display Greek as Beta Code, the word shows up as `ei)/dei+/`, not `ei)/dei/+` as is found in the XML.

U+0301 and U+0308 are not the only combining characters affected. `i)/` normalizes to the single precomposed character U+1F34 GREEK SMALL LETTER IOTA WITH PSILI AND OXIA; while `i/)` becomes the two characters U+03AF GREEK SMALL LETTER IOTA WITH TONOS and U+0313 COMBINING COMMA ABOVE.

This makes me think that it is necessary to reorder Beta Code diacritics before applying them to a Unicode string as combining characters, though I cannot find any references saying something to that effect. All the Unicode diacritics mentioned above have the same canonical combining class, 230, so it doesn't have to do with that.

From what I have shown above, we have a partial order: `)` < `/` and `+` < `/`. My plan is, whenever a character has multiple diacritics, to sort them according to a defined partial order, raising an error if a relation between two diacritics is undefined. That way, we can figure out what to do with any that we have not considered yet.

**Comment by whoopsedesy** [comment]
This is the before and after of rendering:

![Animation of Sh. 1–6 before and after the change](https://user-images.githubusercontent.com/53617134/109411352-610aff80-7999-11eb-8206-77cec80e06f8.gif)

The change affects 1446 words in the corpus. In some cases, reordering the diacritics helps the lemmatizer find a lemma, for example:
```
-Hom.Hymn,2,48,3,δαί̈δας,δαί̈δας,6,--+,auto,1
+Hom.Hymn,2,48,3,δαΐδας,δαίς,6,--+,auto,1
```
```
-Hom.Hymn,2,96,6,δαί̈φρονος,δαί̈φρονος,6.5,-+--,auto,1
+Hom.Hymn,2,96,6,δαΐφρονος,δαίφρων,6.5,-+--,auto,1
```
```
-Il.,1,433,7,νηὶ̈,νηὶ̈,9,+-,auto,1
+Il.,1,433,7,νηῒ,ναῦς,9,+-,auto,1
```

The most commonly affected words are:

|word|count|
|:-----|------:|
|ἀτρεΐδης|77|
|ὀΐω|57|
|νηῒ|50|
|ἀτρεΐδη|46|
|χροῒ|41|
|ἐῢ|34|
|ἠΰτε|31|
|δαΐφρονος|31|
|τυδεΐδης|30|
|ἀτρεΐδαο|26|
|πηλεΐωνα|25|
|πηλεΐδης|24|
|προΐει|21|
|ἀΐδαο|21|
|ὀΐομαι|20|
|διῒ|17|
|ἀΰσας|16|
|ἀτρεΐδην|15|
|τυδεΐδῃ|14|
|πηλεΐωνος|14|

**Comment by whoopsedesy** [comment]
> Indeed, if you configure the Perseus display preferences to display Greek as Beta Code, the word shows up as `ei)/dei+/`, not `ei)/dei/+` as is found in the XML.

tl;dr: The third-party package Hopper uses to convert to/from Beta Code does in fact define a canonical ordering of Greek accents that is congruent to what we now use. I did not find any notes or a rational, so I do not know if it is by design that it correctly handles normalization in the cases we have identified, or if it is a coincidence.

The [latest Hopper source code I could find](https://sourceforge.net/projects/perseus-hopper/files/perseus-hopper/) is unfortunately 10 years old and [no longer supported](https://www.perseus.tufts.edu/hopper/opensource/download) 😔 but we can [see](https://github.com/jfinkels/hopper/blob/b9183dcd8ff74e13e0841fb5bc4d99600395bb06/reading/src/main/java/perseus/document/TextEncodingConverter.java#L20) that it uses an external package, edu.unc.epidoc.transcoder from [EpiDoc](https://sourceforge.net/p/epidoc/wiki/Home/), to do its text conversion.

As I understand it, the EpiDoc transcoder works by converting its input into an internal format consisting of character names with attached diacritic modifiers, like `iota_diaer_oxy` for ΐ. It then converts from this internal format to a desired output format.

The `BetaCodeParser` class, when it is processing the diacritics for a character, [looks up](https://sourceforge.net/p/epidoc/code/2490/tree/trunk/transcoder/src/main/java/edu/unc/epidoc/transcoder/BetaCodeParser.java#l87) the diacritics in a table called [GreekAccents.properties](https://sourceforge.net/p/epidoc/code/2490/tree/trunk/transcoder/src/main/resources/GreekAccents.properties) that maps each to a distinct integer, then attaches the diacritics to the base character in that defined order. Here is a comparison of the two orderings. The differences are that the Sedes one has an entry for `?` (which stands for COMBINING DOT BELOW and which does not appear in the corpus anymore since #11), and that Sedes doesn't enforce an order within the sets {`)`, `(`, `+`} and {`/`, `\`, `=`} (and indeed raises an error if two diacritics from one of those sets are attached to the same character).

<table>
<tr><th>Sedes order</th><th>EpiDoc order</th></tr>
<tr><td><code>?</code></td><td></td></tr>
<tr><td rowspan=3><code>)</code>, <code>(</code>, <code>+</code></td><td><code>)</code></td></tr>
<tr><td><code>(</code></td></tr>
<tr><td><code>+</code></td></tr>
<tr><td rowspan=3><code>/</code>, <code>\</code>, <code>=</code></td><td><code>/</code></td></tr>
<tr><td><code>\</code></td></tr>
<tr><td><code>=</code></td></tr>
<tr><td><code>|</code><td><code>|</code></td></tr>
</table>

The oldest version of GreekAccents.properties in the history is from [2003](https://sourceforge.net/p/epidoc/code/955/log/?path=/trunk/transcoder/src/edu/unc/epidoc/transcoder/GreekAccents.properties), and I don't find anything that explains why the ordering is the way it is.
